src/app/clusters/mode-base-server/mode-base-server.h

```diff
@@ -0,0 +1,215 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "mode-base-cluster-objects.h"
+#include <app/AttributeAccessInterface.h>
+#include <app/AttributePersistenceProvider.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/util/af.h>
+#include <map>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ModeBase {
+
+class Instance : public CommandHandlerInterface, public AttributeAccessInterface
+{
+public:
+    /**
+     * This is a helper function to build a mode option structure. It takes the label/name of the mode,
+     * the value of the mode and a list of mode tags that apply to this mode.
+     *
+     * **NOTE** The caller must ensure that the lifetime of the label and modeTags is as long as the
+     * returned structure.
+     */
+    static chip::app::Clusters::detail::Structs::ModeOptionStruct::Type
+    BuildModeOptionStruct(const char * label, uint8_t mode,
+                          const DataModel::List<const chip::app::Clusters::detail::Structs::ModeTagStruct::Type> modeTags)
+    {
+        chip::app::Clusters::detail::Structs::ModeOptionStruct::Type option;
+        option.label    = CharSpan::fromCharString(label);
+        option.mode     = mode;
+        option.modeTags = modeTags;
+        return option;
+    }
+
+    /**
+     * Initialise the ModeBase server instance.
+     * @return Returns an error if the cluster ID given is not af a valid ModeBase cluster, if the
+     * given endpoint and cluster ID have not been enabled in zap, if the CommandHandler or
+     * AttributeHandler registration fails or if the AppInit() returns an error.
+     */
+    CHIP_ERROR Init();
+
+    // CommandHandlerInterface
+    void InvokeCommand(HandlerContext & ctx) override;
+    CHIP_ERROR EnumerateAcceptedCommands(const ConcreteClusterPath & cluster, CommandIdCallback callback, void * context) override;
+
+    // AttributeAccessInterface
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+    CHIP_ERROR Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder) override;
+
+    // Attribute setters
```

Comment:
These should be documented as storing things in non-volatile storage.

Reply:
Done